Fn: Into<Signal>

With the 0.5.0 changes to the reactive system, it's now easier than ever to convert a `Fn` into a `Signal`, just call `.into_signal()` on it. However, I think `.into()` would be even better. If `Fn` implemented `Into<Signal<_>>`, then those `Fn` could be passed directly to any components taking a `#[prop(into)] Signal<T>`.

`Fn: Into<MaybeSignal<_>>` probably makes sense too. Most of my components currently take `#[prop(into)] MaybeSignal<T>` so that call sites can decide whether to pass in a signal or a function.

As far as I know this is impossible if signals can implement `Fn()`, because it conflicts with the global trait `impl From<T> for T`, as far as I know.

If you try to 
```rust
impl<T, F> From<F> for Signal<T> where F: Fn() -> T {}
```
you get a compiler error that says exactly that
```
error[E0119]: conflicting implementations of trait `From<signal_wrappers_read::Signal<_>>` for type `signal_wrappers_read::Signal<_>`
  --> /Users/gjohnston/Documents/Projects/leptos-main/leptos/leptos_reactive/src/signal_wrappers_read.rs:85:1
   |
85 | impl<T, F> From<F> for Signal<T> where F: Fn() -> T {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T> From<T> for T;

For more information about this error, try `rustc --explain E0119`.
```

> Given that implementing `Fn() -> T` for `Signal<T>` requires nightly, couldn't we use `#![feature(min_specialization)]` (see [specialization](https://github.com/rust-lang/rust/issues/31844))? 

Adding `min_specialization` and/or `specialization` and trying to implement `From` here yields the same error
```
error[E0119]: conflicting implementations of trait `From<signal_wrappers_read::Signal<_>>` for type `signal_wrappers_read::Signal<_>`
  --> leptos_reactive/src/signal_wrappers_read.rs:38:1
   |
38 | impl<F, T> From<F> for Signal<T>
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T> From<T> for T;
```

--- a/leptos_reactive/src/lib.rs
+++ b/leptos_reactive/src/lib.rs
@@ -3,6 +3,8 @@
 #![cfg_attr(feature = "nightly", feature(fn_traits))]
 #![cfg_attr(feature = "nightly", feature(unboxed_closures))]
 #![cfg_attr(feature = "nightly", feature(type_name_of_val))]
+#![cfg_attr(feature = "nightly", feature(auto_traits))]
+#![cfg_attr(feature = "nightly", feature(negative_impls))]
 
 //! The reactive system for the [Leptos](https://docs.rs/leptos/latest/leptos/) Web framework.
 //!

--- a/leptos_reactive/src/signal_wrappers_read.rs
+++ b/leptos_reactive/src/signal_wrappers_read.rs
@@ -428,25 +428,6 @@ impl<T> From<Memo<T>> for Signal<T> {
     }
 }
 
-impl<T: Clone> From<T> for Signal<T> {
-    #[track_caller]
-    fn from(value: T) -> Self {
-        Self {
-            inner: SignalTypes::DerivedSignal(store_value(Box::new(
-                move || value.clone(),
-            ))),
-            #[cfg(any(debug_assertions, feature = "ssr"))]
-            defined_at: std::panic::Location::caller(),
-        }
-    }
-}
-
-impl From<&str> for Signal<String> {
-    fn from(value: &str) -> Self {
-        Self::from(value.to_string())
-    }
-}
-
 enum SignalTypes<T>
 where
     T: 'static,
@@ -851,6 +832,36 @@ impl<T> From<Signal<T>> for MaybeSignal<T> {
 impl From<&str> for MaybeSignal<String> {
     fn from(value: &str) -> Self {
         Self::Static(value.to_string())
+    }
+}
+
+#[cfg(feature = "nightly")]
+mod from_fn_for_signals {
+    use super::{MaybeSignal, Memo, ReadSignal, RwSignal, Signal};
+    auto trait NotSignalMarker {}
+
+    impl<T> !NotSignalMarker for Signal<T> {}
+    impl<T> !NotSignalMarker for ReadSignal<T> {}
+    impl<T> !NotSignalMarker for Memo<T> {}
+    impl<T> !NotSignalMarker for RwSignal<T> {}
+    impl<T> !NotSignalMarker for MaybeSignal<T> {}
+
+    impl<F, T> From<F> for Signal<T>
+    where
+        F: Fn() -> T + NotSignalMarker + 'static,
+    {
+        fn from(value: F) -> Self {
+            Signal::derive(value)
+        }
+    }
+}
+#[cfg(not(feature = "nightly"))]
+impl<F, T> From<F> for Signal<T>
+where
+    F: Fn() -> T + 'static,
+{
+    fn from(value: F) -> Self {
+        Signal::derive(value)
     }
 }
 

